Source/WebCore/Modules/filesystemaccess/FileSystemWritableFileStreamSink.cpp

```diff
@@ -0,0 +1,173 @@
+{
+    WTF::switchOn(data, [&](const RefPtr<JSC::ArrayBufferView>& bufferView) {
+        if (!bufferView || bufferView->isDetached())
+            return completionHandler(false, { });
+
+        RefPtr buffer = bufferView->possiblySharedBuffer();
+        if (!buffer)
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<JSC::ArrayBuffer>& buffer) {
+        if (!buffer || buffer->isDetached())
+            return completionHandler(false, { });
+
+        completionHandler(true, buffer->span());
+    }, [&](const RefPtr<Blob>& blob) {
+        if (!blob)
+            return completionHandler(false, { });
+
+        blob->getArrayBuffer([completionHandler = WTFMove(completionHandler)](auto&& result) mutable {
+            if (result.hasException())
+                return completionHandler(false, { });
+
+            Ref buffer = result.releaseReturnValue();
+            if (buffer->isDetached())
+                return completionHandler(false, { });
+
+            completionHandler(true, buffer->span());
+        });
+    }, [&](const String& string) {
+        completionHandler(true, string.utf8().span());
+    });
+}
+
+ExceptionOr<Ref<FileSystemWritableFileStreamSink>> FileSystemWritableFileStreamSink::create(FileSystemFileHandle& source)
+{
+    return adoptRef(*new FileSystemWritableFileStreamSink(source));
+}
+
+FileSystemWritableFileStreamSink::FileSystemWritableFileStreamSink(FileSystemFileHandle& source)
+    : m_source(source)
+{
+}
+
+FileSystemWritableFileStreamSink::~FileSystemWritableFileStreamSink()
+{
+    if (!m_isClosed)
+        protectedSource()->closeWritable(false);
+}
+
+// https://fs.spec.whatwg.org/#write-a-chunk
+void FileSystemWritableFileStreamSink::write(ScriptExecutionContext& context, JSC::JSValue value, DOMPromiseDeferred<void>&& promise)
+{
+    if (m_isClosed)
+        return promise.reject(Exception { ExceptionCode::InvalidStateError, "Sink is closed"_s });
+
+    if (value.isUndefined())
+        return promise.reject(Exception { ExceptionCode::TypeError, "Data is undefined"_s });
+
+    auto scope = DECLARE_THROW_SCOPE(context.vm());
+    auto chunkResult = convert<IDLUnion<IDLArrayBufferView, IDLArrayBuffer, IDLInterface<Blob>, IDLUSVString, IDLDictionary<FileSystemWritableFileStream::WriteParams>>>(*context.globalObject(), value);
+    if (UNLIKELY(chunkResult.hasException(scope)))
+        return promise.reject(Exception { ExceptionCode::ExistingExceptionError });
+
+    auto writeParams = writeParamsFromChunk(chunkResult.releaseReturnValue());
+    switch (writeParams.type) {
+    case FileSystemWriteCommandType::Seek: {
+        if (!writeParams.position)
+            return promise.reject(Exception { ExceptionCode::TypeError, "Position is not provided for seek"_s });
```

Comment:
According to spec, we are supposed to reject asynchronously since we are going to the file system queue and then to the storage task queue.
This is observable to JS though an edge case.
I would be tempted to change the spec to check for the parameters synchronously before enqueueing tasks.

Ditto below

Reply:
ok I will drop the check to match spec; and file a spec issue.